### ISSUE TITLE
[feature request] Please add support for sending files

### ISSUE THREAD

**Issue by RogueAI42** [george99g/QDiscord#8]
A bot I'm working on generates an image using Magick++ and should send it to the channel that triggered it. However, QDiscord only seems to support sending plaintext messages. I've tried attaching the generated data to a QHttpMultiPart request manually but absolutely nothing happens. It would be very helpful to be able to use the API for this.

**Comment by ArsenArsen** [comment]
This is (to my awareness) going to be available after the refactor/rewrite, so in the near future.

**Comment by george99g** [comment]
> This is (to my awareness) going to be available after the refactor/rewrite, so in the near future.

This is, in fact, currently working with the code I have locally. I'm currently working on it, though. It'll take some time.

**Comment by Swerved** [comment]
I've tried this as well, and struggling to upload files with HttpMultiPart... (about as much info I'm getting is 'unknown error'). What I have as time of writing:

    QHttpPart imagePart;
    QFileInfo fi(filename);
    QFile *file = new QFile(filename);
    imagePart.setBodyDevice(file);
    imagePart.setHeader(QNetworkRequest::ContentTypeHeader, QVariant("image/png"));
    imagePart.setHeader(QNetworkRequest::ContentDispositionHeader, QVariant("form-data; filename=\"image.png\"; name=\"file\""));

Great work on this library but I've already got a very slim minimal function with sending messages to a particular webhook, so I'm happy there. 

If you do have any pointers with this george99g that would be much appreciated, hope you don't mind :)

**Comment by ArsenArsen** [comment]
Here, have a pointer: `0x2b9493194010` :^)
On a more serious note, `name` disposition should be `file`, filename shold be filename, but content type should be `application/octet-stream`. There should also be a part called `payload_json` which contains a JSON representation of the message.

**Comment by Swerved** [comment]
Thanks. The docs seem to suggest json isn't required, how I read it. In any case I seems I should try that because its still failing.
Will probably take a bit of time to do the json bit, so I will try that later.

**Comment by george99g** [comment]
I've currently got it working here with the following code, but this requires a version of the library that is currently only available on my local repo. I've still got a little bit of work to do before it can be used without writing 20 lines of code just for setup.

```cpp
QDiscordRest _drst;
_drst.setToken(QDiscordToken("token"));

QHttpMultiPart* multiPart = new QHttpMultiPart(QHttpMultiPart::FormDataType);
QHttpPart contentType;
contentType.setHeader(QNetworkRequest::ContentDispositionHeader, "form-data; name=\"content\"");
contentType.setBody("test");
multiPart->append(contentType);
QHttpPart filePart;
QFile* file = new QFile("/home/user/myFile.txt");
file->open(QFile::ReadOnly);
filePart.setHeader(QNetworkRequest::ContentTypeHeader, QMimeDatabase().mimeTypeForFile(file->fileName()).name());
filePart.setHeader(QNetworkRequest::ContentDispositionHeader, "form-data; name=\"file\"; filename=\"myfile.txt\"");
filePart.setBodyDevice(file);
file->setParent(multiPart);
multiPart->append(filePart);

QNetworkRequest request;
_drst.request(request, QDiscordRoutes::Messages::sendMessage(QDiscordID(169867030806265856ULL)), multiPart, [multiPart](QNetworkReply* reply)
{
	qDebug()<<reply->errorString();
	delete multiPart;
});
```

**Comment by Swerved** [comment]
Thanks. Obviously I don't have QDiscordRest class, but I'll have a go without it - was unaware of QMimeDatabase, that might actually help me here. 

Look forward to seeing the next update for this, much appreciate your help.

**Comment by george99g** [comment]
You're very welcome! You could join https://discord.gg/MGdYDqC if you want to get support a little bit faster or get status reports every now and then.

**Comment by george99g** [comment]
The code I posted earlier should now work with aca95a01431cc18f7c1febe224da697fc51bb0a5. I'll work on actually supporting attachments later.

**Comment by george99g** [comment]
Implemented in 2ff014975cb2b320520c70bce557b5045a86354c.
